### PR TITLE
Use full URL in `http.url` tag

### PR DESCRIPTION
**Pull request by lloeki**
**What does this PR do?**

*Adjust HTTP URL tags to comply with span tag unification RFC*

- align http.url with tag naming RFC (and Rack's #url) to include the
  complete request URL
- drop http.base_url, now unneeded
- preserve REQUEST_URI logic that differs from Rack's logic
- add SCRIPT_NAME which was previously unhandled for PATH_INFO
- append QUERY_STRING when using PATH_INFO to match REQUEST_URI

*Enable query string quantization by default*
    
This change is mandated by the unifying tag RFC.

**Motivation**

Compliance with internal RFC document.

### CODE REVIEW

**Review by delner — CHANGES_REQUESTED** [review]
Please wait on this pending internal discussion. Thanks!

**Comment by lloeki** [comment]
> Adjust HTTP URL tags to comply with span tag unification RFC

This is now made optional via a quantization option.

> drop http.base_url, now unneeded

This is now only dropped when `http.url` includes the URL base.

> append QUERY_STRING when using PATH_INFO to match REQUEST_URI

This fixes a discrepancy between the happy path and the fallback.

> Enable query string quantization by default

This is now removed from the PR.

**Comment by lloeki** [comment]
Save for one test case which was adjusted for an inconsistency that is now resolved, the behaviour does not change, thus `rake spec:rack` tests pass right away.

I shall now add a couple of test cases to validate the new behaviour when the option is set.

File: spec/datadog/tracing/contrib/sinatra/multi_app_spec.rb
Comment (marcotc) [inline spec/datadog/tracing/contrib/sinatra/multi_app_spec.rb:77]: Could you explain why these Sinatra tests have changed?

File: lib/datadog/tracing/contrib/utils/quantization/http.rb
Comment (marcotc) [inline lib/datadog/tracing/contrib/utils/quantization/http.rb:22]: Can you add unit tests for the changes in this file?

File: spec/datadog/tracing/contrib/sinatra/multi_app_spec.rb
Comment (lloeki) [reply spec/datadog/tracing/contrib/sinatra/multi_app_spec.rb:77]: These tags actually come from Rack, not Sinatra. I believe they're here because at some point Sinatra may have been backfilling this value by itself. In any case they act as an integration test checking for side effects.

Sadly this test would previously pass but not represent what happens when real-world `REQUEST_URI` is set, indeed `Rack::Test` does not set `REQUEST_URI`.

The change here is mainly due to a bug that this PR addresses, that previously made the fallback to `PATH_INFO` be inconsistent with `REQUEST_URI` (i.e it did not include `SCRIPT_NAME` nor `QUERY_STRING`), which created all kinds of special cases from the lack of query string to adding `base_url tag + url tag` not resulting in the same value depending on whether `REQUEST_URI` or `PATH_INFO` was used, not the least in the latter case `base_url tag + url tag` is entirely nonsensical for virtualhosted or nested apps since it misses the intermediate `SCRIPT_NAME`.

File: lib/datadog/tracing/contrib/utils/quantization/http.rb
Comment (lloeki) [reply lib/datadog/tracing/contrib/utils/quantization/http.rb:22]: Yup, I have them right here, I just detected a last second issue that needed some adjustment.

File: docs/GettingStarted.md
Comment (lloeki) [inline docs/GettingStarted.md:1538]: Ironically, the doc here was consistent with the internal `Quantization::HTTP.url` but not with the Rack integration `http.url` tagging behaviour!

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (lloeki) [inline lib/datadog/tracing/contrib/rack/middlewares.rb:27]: I can get the argument for method length (with exceptions), but I'm wondering if there's really such a thing as a class with "too much lines".

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (lloeki) [inline lib/datadog/tracing/contrib/rack/middlewares.rb:175]: This has to be done here, as doing it as a config default would be silently overridden when `:quantize` is manually set, which would be most surprising to the user.

Probably a general problem with such nested configuration keys.

File: spec/datadog/tracing/contrib/utils/quantization/http_spec.rb
Comment (lloeki) [inline spec/datadog/tracing/contrib/utils/quantization/http_spec.rb:50]: This seemed to be a typo/eager copy-pasting.

**Comment by lloeki** [comment]
Hmm there's flakiness still:

```
# pass
bundle exec appraisal ruby-3.1.1-contrib rake spec:rack['--seed 39851']
# fail
bundle exec appraisal ruby-3.1.1-contrib rake spec:rack['--seed 17324']
```

Had one like that before, seems like `quantize:` config is leaking from one spec example to the next.

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (marcotc) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:27]: I don't think I've ever decided to break down a class when `Metrics/ClassLength` is triggered, only ever added the cop exclusion tag.

If you want to remove the cop, I'd ✅ it.

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (marcotc) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:175]: If I understand it correctly, this will be removed in the next major release, is that right?

If so, I suggest writing them down in the comment, so we can find it later.

Comment (marcotc) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:175]: Actually, can you create a new file named `docs/NextReleaseUpgradeGuide.md` (name suggestion only), just add one section:
```
# Breaking changes

To aid with the next release process, please tag future breaking changes the inline code comment `# DEV-2.0:` prefix. 

## List of breaking changes

* `http.url` will change its default...
```

We've been meaning to do it, and it seems like your PR wants to have a breaking change in the next release.

Comment (marcotc) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:175]: With this, I suggest adding the comment in this line with the prefix `# DEV-2.0:`.

File: docs/GettingStarted.md
Comment (lloeki) [reply docs/GettingStarted.md:1553]: You mean the updated comments? I understand the confusion, as this is actually not related to this change: the current default is that the `http.url` tag contains a path, not a full URL.

Indeed `Quantization::HTTP.url` does not care whether it gets passed a path or a full URL, and the Rack integration merely used to pass a path to be quantized. Now the Rack integration passes a full URL and defaults to telling `Quantization::HTTP.url` to remove the URL base in order to preserve the previous behaviour.

So in essence the comments were technically correct regarding what the internal `Quantization::HTTP.url` method does - which, I guess, is why it was documented this way - but incorrect as to what the actual tagging behaviour by the Rack integration was doing previously and now does as a default, which is what this doc refers to, and what the user wants to know about.

Comment (marcotc) [reply docs/GettingStarted.md:1553]: Thanks for clarifying.

**Review by marcotc — APPROVED** [review]
Looks 👍 after the latest changes.

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (lloeki) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:175]: Good points! Will do.

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (lloeki) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:27]: Seems like there are not too many of them, so it can reasonably fit into this PR.

File: lib/datadog/tracing/contrib/rack/middlewares.rb
Comment (lloeki) [reply lib/datadog/tracing/contrib/rack/middlewares.rb:175]: I think I'll do that in another PR, along with this suggestion: https://github.com/DataDog/dd-trace-rb/pull/2283#discussion_r975902685

**Comment by lloeki** [comment]
Alright, all comments addressed (thanks @marcotc!), waiting for CI.

@delner I'll proceed by dismissing your review since the concern has been discussed internally and addressed.

**Comment by lloeki** [comment]
All good, merging 🎉

**Comment by gingerlime** [comment]
Hi @lloeki. Thanks for the fix. I'm testing dd-trace 1.5.1 and still see the same issue unfortunately.

```
root@host:/var/local/app/app/dependabot/bundler/ddtrace-1.5.1 %)# curl "http://localhost:3000/|"

Puma caught this error: bad URI(is not URI?): "/|" (URI::InvalidURIError)
/usr/local/lib/ruby/3.0.0/uri/rfc3986_parser.rb:67:in `split'
/usr/local/lib/ruby/3.0.0/uri/rfc3986_parser.rb:72:in `parse'
/usr/local/lib/ruby/3.0.0/uri/rfc3986_parser.rb:106:in `convert_to_uri'
/usr/local/lib/ruby/3.0.0/uri/generic.rb:1100:in `merge'
/usr/local/lib/ruby/3.0.0/uri/rfc3986_parser.rb:78:in `inject'
/usr/local/lib/ruby/3.0.0/uri/rfc3986_parser.rb:78:in `join'
/usr/local/lib/ruby/3.0.0/uri/common.rb:208:in `join'
/usr/local/bundle/gems/ddtrace-1.5.1/lib/datadog/tracing/contrib/rack/middlewares.rb:274:in `parse_url'
/usr/local/bundle/gems/ddtrace-1.5.1/lib/datadog/tracing/contrib/rack/middlewares.rb:167:in `set_request_tags!'
/usr/local/bundle/gems/ddtrace-1.5.1/lib/datadog/tracing/contrib/rack/middlewares.rb:110:in `call'
/usr/local/bundle/gems/rack-mini-profiler-3.0.0/lib/mini_profiler/profiler.rb:249:in `call'
/usr/local/bundle/gems/railties-7.0.4/lib/rails/engine.rb:530:in `call'
/usr/local/bundle/gems/puma-5.6.5/lib/puma/configuration.rb:252:in `call'
/usr/local/bundle/gems/puma-5.6.5/lib/puma/request.rb:77:in `block in handle_request'
/usr/local/bundle/gems/puma-5.6.5/lib/puma/thread_pool.rb:340:in `with_force_shutdown'
/usr/local/bundle/gems/puma-5.6.5/lib/puma/request.rb:76:in `handle_request'
/usr/local/bundle/gems/puma-5.6.5/lib/puma/server.rb:443:in `process_client'
/usr/local/bundle/gems/puma-5.6.5/lib/puma/thread_pool.rb:147:in `block in spawn_thread'
```

**Comment by gingerlime** [comment]
@lloeki yeah, sorry, I wasn't sure where to post this. I already linked to this comment on the other PR though.

**Pull request merged**
